power_cell_detector.py

```diff
@@ -0,0 +1,43 @@
+import cv2
+import numpy as np
+
+
+class PowerCellDetector:
+    def __init__(self, generator):
+        self.input = generator
+        self.LOW_YELLOW = np.array([17, 120, 50])
+        self.HIGH_YELLOW = np.array([33, 255, 255])
+
+    # Returns a tuple with:
+    # - a list of detected circles, each with [x-coordinate, y-coordinate, radius].
+    # - the depth frame
+    def run_detector(self):
+        img, depth_frame = self.input.generate()
+
+        if img is None:
+            return None, None
+
+        hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
+        hsv = cv2.medianBlur(hsv, 5)
+        mask = cv2.inRange(hsv, self.LOW_YELLOW, self.HIGH_YELLOW)
+
+        # Erode removes white noise at edges of the contours.
+        mask = cv2.erode(mask, None, iterations=7)
+
+        # Dilation increases the overall area of the object to accentuate the contours.
+        mask = cv2.dilate(mask, None, iterations=10)
+        output = (mask == 255) * hsv[:, :, 1]
+
+        # The second-to-last parameter is the block size, which is the width of the square region to use to calculate
+        # the threshold for a pixel--the threshold is different in different regions of the frame.
+        th = cv2.adaptiveThreshold(output, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, cv2.THRESH_BINARY_INV, 15, 10)
```

Comment:
How much benefit are we getting from doing this second layer of thresholding?

Reply:
I tested this actually. It makes a significant difference actually. It returns a lot of false balls along with the correct ones.